### ISSUE TITLE
To shell or not to shell ?

### ISSUE THREAD

**Issue by roadelou** [theodumont/video-logging#9]
This is more of an __architectural decision__ than a code issue, but I think it is important to bring it up regardless. What is at stake here is that the application you wrote is behaving __kind of like a shell__, but only implements __a subset of a shell capabilities__.

This brings several issues :
 - A __lot of commands are not available__ in you shell, for instance you did not write a 'ls' command so I cannot see the content of a folder before sorting it. In fact, no tool from the outside is available within the shell.
 - Some __cool features__ that are built into shells are __not available here__ (up arrow to use last command, tab to autocomplete etc...).
 - For a third party, it is __not possible to extend__ what the tool does (because you cannot add a new command without changing the source code).
 - It is __not easy to interface__ your tool with other existing tools, you would have to feed it its command with echo for instance.
 - There is __no programming language__ within your shell, therefore you cannot 'script' with it.

The risk with this approach is that you might end up in some kind of an __uncanny valley__, where your tool is doing its own thing so it is hard to use it within other tools, but at the same time it is not flexible enough to stand on its own, and therefore you still need to go back to the shell when you are working.

To solve this issue, I would advise you to __go one of the two ways__ : either make this tool a __CLI__ thing that can be integrated within the existing shell, or make it an entirely __independent shell__.

# The CLI way

Turning your project into a very convenient CLI tool should not be very difficult, in fact you would __only need a few things__ (not necessarily all at once !)

### Make it a python package

By turning your tool into a python package, you can set its "entry point". For instance, when you install the black code formatter in python, it will create a short script (called 'black') that calls the right function within the black library. You can then use it very conveniently from your shell.

### Use argv and not just input

The command line arguments are received from argv and not from stdin. If you were to process argv and then switch to stdin it would let people use you code as (for instance) :

```bash
video-logging folder
video-logging trash 60
video-logging date
video-logging help
```

### Follow GNU style CLI guidelines

A lot of Windows command line tools use the dos style CLI (I thing it comes from dos) :
```sh
# On windows, using cmd.exe
dir /s/w/o/p/a:-d
```

But that is not the one we are used to, and most tools use the GNU style format derived from getopt :

```sh
ls -a -F
```

For commands that are actually a proxy to other commands, (as is your case), one can use a 'git-style' CLI :

```sh
# Not how "-v" behaves the same way for commit and push
git commit -v -a -m "Commit message"
git push -v
```

It is not recommended that you try to parse the GNU style command line yourself, instead you could use something like [argparse](https://docs.python.org/3/library/argparse.html) in python. This way, one could use your tool as (for example) :

```bash
# To sort the video files (-e for --extension)
video-logging folder -e Videos
# -t for --time for instance
video-logging trash -t 60
video-logging --help
video-logging date --help
```

Just to be clear, this would mean that you have to use argparse once for the global video-logging tool and once for every sub tool (folder, trash etc...) if you follow this example.

By the way, it is advised to __let the user to set the verbosity__ level of a tool from the command line with flags such as :

```bash
video-logging -v folder
video-logging --silent trash -t 60
# Or if you are on the quiet team
video-logging --quiet date
```

### Don't give up on the nice interactive user interface

One simple way to keep the interface of video-logging (which is quite nice by the way :heart:) is to take advantage of the symmetry between argv and stdin (if you decide to implement it).

That would require a bit of fiddling with argparse, but it clearly able to handle it, you just have to tell it to __stop parsing at the first non recognized argument__. That may very well be the default behavior of argparse by the way.

You could for instance decide that the user interface will start if video-logging is __called without a sub program name__ :

```bash
# This one would start the interaction
video-logging
# This one would also start the interaction
video-logging --verbose
# This one is debatable, it depends on your interpretation
video-logging -s # -s for --silent
```

This way your project can be fully taken advantage of as a CLI tool in scripts, but a human user could also use the nice interface.

Note that the CLI I presented is __just an example__, there are countless other ways to do this, just find one that you are comfortable with.

# The shell way

Before I go any further, I have to emphasis that this would __take a significant amount of work__ :hourglass: and I am not completely sure how this would concretely be achieved.

The main issue when writing a shell is that it must __not only support a lot of built-in__ utilities (like cd, the pipe '|', redirection '>' etc...), but it also has to __embed an entire programming language__ to be usable in real world scenarios. As you already know, __writing a programming language is very hard__ on its own, but writing one for the command line holds an __additional complexity__.

Indeed, a shell language must not only support standard language constructs :

```bash
# That is a function in bash, other shell can have different a syntax
function hello_world {
  message="Hello, World!"
  echo $message
  return $message
}
```

But it must also support one liners that are commonly used :

```bash
grep "Hello, World!" ~/ISSUE.md | grep -v $EXCLUDE > hello_world.txt
```

The syntax of the shell language must be able to make sense of both code snippets, which leads to non-trivial (as in _extremely complex_) syntaxes.

I guess that this warning should be enough to push you toward the first decision :wink: (if you ever considered the second one), but in case it did not I will try to share __some hints on how that could be achieved__.

Fortunately, you might not even have to write the shell syntax yourself. There are __already a lot of usable implementations__ out there. I believe that the most common embedded shell-like interpreter is [TCL](https://www.tcl.tk/).

The way this would work is through __extensions__. TCL for instance allows you to extend its core with extensions written in some language (I will blindly guess C/C++). Writing such extensions probably is fairly challenging, but I have never done it myself so __don't take my word for it__.

On the bright side, you now (or in a month of work to be fair :smile:) have a fully functional shell that has some nice video handling abilities. However, you had to __rewrite all of your code to C/C++/something else__, which might not be what you were looking for.

# Conclusion

To sum things up :
 - If you are just __trying to create a tool__ to perform various video-related manipulations, you probably could tweak your implementation to make it fit nicely in the __CLI__ world.
 - If you are looking to __build a fully fledged shell__, I would advise that you first __find a preexisting shell, and then extend it__ to perform video manipulations.

__The choice is yours__, and it only depends on what you want to do with your project (and how much time you have ahead of you :laughing:). I would however like to emphasize that __making no choice here would probably be worse__ than any of the two paths I described :smile:

---

EDIT: After verification, I was wrong about git-style CLI, I changed the issue to the correct version :wink:
